wgsl/index.bs

```diff
@@ -4081,6 +4081,72 @@ TODO: user data inputs of fragment stage must be subset of user data outputs of
 
 ### Input-output matching rules TODO ### {#input-output-matching}
 
+# Language extensions # {#language-extensions}
+
+The [SHORTNAME] language is expected to evolve over time.
+
+An <dfn noexport>extension</dfn> is a named grouping for a coherent
+set of modifications to a particular version of the [SHORTNAME] specification, consisting of any combination of:
+* Addition of new concepts and behaviours via new syntax, including:
+    * declarations, statements, attributes, and built-in functions.
+* Removal of restrictions in the current specification.
```

Comment:
What about modifications of the shading language that want to *add* new restrictions? I'm thinking WebGPU-compat, where OpenGL ES or D3D11 may not like some of the things we have in baseline WGSL.

Reply:
I want "enable" to be clearly additive (or "max").  If a code reader quickly scans the source, the union of used features is at most the union of the enables.
Also, it aids composability: I can glue together source code together.

Otherwise you get a see-saw battle in adding and subtracting features, which can be error prone. (E.g. like resolving positive and negative ACLs.)

I could go with an "enable foo;", where foo then enables a new *syntax* that limits functionality for a particular module-scope declaration.

First example:

       // Let's say we targeted an API that had did not support atomic operations.
       // Declare that we'll use "no_atomics" attribute on entry point functions.
       enable compat_no_atomics;

       // Adding no_atomics says this shader and its static call tree must not use
       // atomic types (or perform any atomic accesses).
       [[stage(compute), no_atomics]] 
       fn mycompute(...) -> void { 
           ...
       }

Some external convention or tool would have to enforce the rule that all entry points were annotated with `no_atomics`.

Alternatively, in this second example, the feature that is enabled is a *different* module-scope declaration to disable a feature _for the rest of the module_.    This isn't very composable, but it requires less writing.

       // Let's say we targeted an API that had did not support atomic operations.
       // Declare that we might use a module-scope 'disable atomics'.
       enable compat_no_atomics;

       // Declare that nowhere in the rest of the source module will it be valid to
       // use atomic types, or perform atomic accesses.
       disable atomics;

       [[stage(compute)]] 
       fn mycompute(...) -> void { 
           ...
       }